File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Combination Sum II/Combination Sum II.cpp
class Solution {
private:
    void myHelper(vector<vector<int>> &res, vector<int> &help, vector<int> &candidates, int target, int currSum, int ind) {
        if (currSum == target) {
            res.push_back(help);
            return; // Stop further recursion once target is reached
        }
        if (currSum > target || ind >= candidates.size()) {
            return;
        }

        for (int i = ind; i < candidates.size(); ++i) {
            if (i > ind && candidates[i] == candidates[i - 1]) {
                continue; // Skip duplicates
            }
            
            help.push_back(candidates[i]);
            myHelper(res, help, candidates, target, currSum + candidates[i], i + 1);
            help.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> help;
        sort(candidates.begin(), candidates.end()); // Sort to handle duplicates
        myHelper(res, help, candidates, target, 0, 0);
        return res;
    }
};

File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Combination Sum II/Combination Sum II.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] arr, int target) {
        Arrays.sort(arr);
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();
        combi(0,arr,target,ans,ds);
        return ans;
    }
    public void combi(int id,int arr[],int target,List<List<Integer>> ans,List<Integer> ds)
    {
        if(target==0)
        {
            ans.add(new ArrayList<>(ds));
        }
        for(int i=id;i<arr.length;i++)
        {
            if(i>id && arr[i]==arr[i-1])
            {
                continue;
            }
            if(arr[i] > target)
            {
                break;
            }
            ds.add(arr[i]);
            combi(i+1,arr,target-arr[i],ans,ds);
            ds.remove(ds.size()-1);
        }
        
    }
}